win32/src/file_win32.c

```diff
@@ -60,4 +41,472 @@ static void on_file_io_complete_win32(
+                    /*Codes_SRS_FILE_WIN32_43_003: [ file_create shall initialize a threadpool environment by calling InitializeThreadpolEnvironment.]*/
+                    InitializeThreadpoolEnvironment(&handle->cbe);
+
+                    /*Codes_SRS_FILE_WIN32_43_004: [ file_create shall obtain a PTP_POOL struct by calling execution_engine_win32_get_threadpool on execution_engine.]*/
+                    handle->ptp_pool = execution_engine_win32_get_threadpool(execution_engine);
+
+                    /*Codes_SRS_FILE_WIN32_43_005: [ file_create shall register the threadpool environment by calling SetThreadpoolCallbackPool on the initialized threadpool environment and the obtained ptp_pool ]*/
+                    SetThreadpoolCallbackPool(&handle->cbe, handle->ptp_pool);
+
+                    /*Codes_SRS_FILE_WIN32_43_006: [ file_create shall create a cleanup group by calling CreateThreadpoolCleanupGroup.]*/
+                    handle->ptp_cleanup_group = CreateThreadpoolCleanupGroup();
+
+                    if (handle->ptp_cleanup_group == NULL)
+                    {
+                        LogError("Failure in CreateThreadpoolCleanupGroup, full_file_name=%s", full_file_name);
+                        result = false;
+                    }
+                    else
+                    {
+                        /*Codes_SRS_FILE_WIN32_43_007: [ file_create shall register the cleanup group with the threadpool environment by calling SetThreadpoolCallbackCleanupGroup.]*/
+                        SetThreadpoolCallbackCleanupGroup(&handle->cbe, handle->ptp_cleanup_group, on_close_threadpool_group_member);
+
+                        /*Codes_SRS_FILE_WIN32_43_033: [ file_create shall create a threadpool io with the allocated FILE_HANDLE and on_file_io_complete_win32 as a callback by calling CreateThreadpoolIo]*/
+                        handle->ptp_io = CreateThreadpoolIo(handle->h_file, on_file_io_complete_win32, NULL, &handle->cbe);
+
+                        if (handle->ptp_io == NULL)
+                        {
+                            LogError("Failure in CreateThreadpoolIo, full_file_name=%s", full_file_name);
+                            result = false;
+                        }
+                        if (!result)
+                        {
+                            CloseThreadpoolCleanupGroup(handle->ptp_cleanup_group);
+                        }
+                    }
+                    if (!result)
+                    {
+                        DestroyThreadpoolEnvironment(&handle->cbe);
+                    }
+                }
+            }
+            /*Codes_SRS_FILE_43_034: [ If there are any failures, file_create shall fail and return NULL. ]*/
+            if (!result)
+            {
+                free(handle);
+            }
+        }
+    }
+    /*Codes_SRS_FILE_WIN32_43_008: [ If there are any failures, file_create shall return NULL.]*/
+    if (!result)
+    {
+        return NULL;

```

Comment:
>return NULL; [](start = 8, length = 12)

style: Should only have one return at the end of the function. Can use "goto" if it makes error handling/cleanup easier, but there should be one variable (likely called result) that is set for each failure or success branch and returned at the end of the function.